### ISSUE TITLE
[1.10] Volume plugin API broken?

### ISSUE THREAD

**Issue by holgerreif** [docker/docker#21857]
**Output of `docker version`:**
```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**
```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 9
Server Version: 1.10.3
Storage Driver: btrfs
 Build Version: Btrfs v3.16+20140829
 Library Version: 101
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local nfs cifs
 Network: null host bridge
Kernel Version: 3.12.53-60.30-default
Operating System: SUSE Linux Enterprise Server 12 SP1
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 5.728 GiB
Name: SOMENAME
ID: KTVH:NB4H:RWZH:7MEA:ZYBF:EIHE:GFLD:T7ZO:47Z7:KWSK:NZCE:SKU4
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

It is a virtual machine

**Steps to reproduce the issue:**
1. Install volume plugin [docker-volume-netshare](https://github.com/gondor/docker-volume-netshare) at least v 0.14
2. Start it with follwing otpions: `docker-volume-netshare nfs -v 3 --verbose=true`
3. `docker volume create -d nfs --name host/share2 --opt blah=blub`
4. Find in messages `Entering Get: {host/share2 map[]}` produced by `docker-volume-netshare`
5. `docker volume ls`
6. Don't finde the created volume listed
7. `docker run -it -v  host/share2:/mnt --rm  some/image /bin/bash`
8. Find the volume mounted
9. `docker volume ls`
10. Find the nfs-volume listed

**Describe the results you received:**
```
docker-volume-netshare: level=debug msg="Entering Get: {host/share2 map[]}"
docker-volume-netshare: level=debug msg="Entering List: { map[]}"
```
**Describe the results you expected:**
With docker 1.9.1 the following sequence of calls to the plugin-API occured:
```
docker: level=info msg="POST /v1.21/volumes/create"
docker-volume-netshare: level=debug msg="Entering Create: name: host/share, options map[blah:blub]"
docker-volume-netshare: level=debug msg="Create volume -> name: host/share, map[blah:blub]"
docker-volume-netshare: level=debug msg="Entering Path: {host/share map[]}"
docker-volume-netshare: level=debug msg="Path for host/share is at /var/lib/docker-volumes/netshare/nfs/host/share"
docker-volume-netshare: level=debug msg="Entering List: { map[]}"
```

**Additional information you deem important (e.g. issue happens only occasionally):**
The only difference between expected and received results is the different docker version.
So either the plugin api changed in some way or we hit a bug. 

It seems impossible to hand over options from `volume create` to the plugin.

I havn't tested with other volume plugins.

This bug was discovered in [this issue](https://github.com/gondor/docker-volume-netshare/issues/21)

**Comment by thaJeztah** [comment]
ping @cpuguy83 could you have a look at this?

**Comment by cpuguy83** [comment]
It looks like the `List` call for netshare did not return any values, and as such `docker volume ls` will not list anything.

**Comment by thaJeztah** [comment]
@cpuguy83 I think the problem/question is why "create" is never called in the steps described, whereas on docker 1.9 it is called

**Comment by holgerreif** [comment]
@cpuguy83 you are right. Plugin's list call returns no share.However the reason for this is the plugin not receiving a create call as in version 1.9.1.
This is what i treat as bug.
Once the volume is used once in a docker run call, it is shown in the list as expected.


--Viele GrüßeHolger Reif

-------- Ursprüngliche Nachricht --------
Von: Brian Goff <notifications@github.com> 
Datum: 07.04.2016  20:46  (GMT+01:00) 
An: docker/docker <docker@noreply.github.com> 
Cc: Holger Reif <holger-git@reif.net> 
Betreff: Re: [docker/docker] [1.10] Volume plugin API broken? (#21857) 

It looks like the List call for netshare did not return any values, and as such docker volume ls will not list anything.

**Comment by cpuguy83** [comment]
@holgerreif Right, a `Get` is performed to see if the volume already exists on the driver.
This behavior should not affect the `List` operation.
Note that both `List` and `Get` calls were added in 1.10.

**Comment by holgerreif** [comment]
The only explaination for the observed behaviour is that the volume plugin returns something non-empty for the Get-Call and thus docker engine does not call Create.
@cpuguy83 could you confirm?

I need to understand the plugin's code what is returning something non-empty

**Comment by cpuguy83** [comment]
@holgerreif Correct `Create` will not be called if `Get` returned something.

**Comment by holgerreif** [comment]
Thanks, I think you can close the issue.

BTW: [plugins.md](https://github.com/docker/docker/blob/master/docs/extend/plugins_volume.md) lists the operations at the API. Is there somewhere more information (besides code ;-) about the sequence (e.g. `Create`conditionally called after `Get` for `docker volume create`)?

**Comment by thaJeztah** [comment]
@holgerreif there's this repository; https://github.com/docker/go-plugins-helpers for people developing a plugin.

For this case; perhaps you have suggestions to improve the https://github.com/docker/docker/blob/master/docs/extend/plugin_api.md documentation? Are you interested in opening a PR to make this more clear?

**Comment by holgerreif** [comment]
My idea was something along this lines: which action on the remote api side (expressed by docker client calls) will trigger which action on the backend handler side. Do you think this might be helpfull?

I'll see, whether I can improve https://github.com/docker/docker/blob/master/docs/extend/plugins_volume.md somehow (and get it even correct ;-)

**Comment by thaJeztah** [comment]
Thanks! At least we should emphasize that *anything* that's returned by `Get` results in the `Create` being skipped (if I understand correctly - I'm not fully up to speed with the plugin API)

**Comment by HackToday** [comment]
If volume doc could be improved like https://github.com/docker/docker/blob/master/docs/extend/plugins_authorization.md

it would be helpful, right now, volume plugin help doc not contain much info

**Comment by holgerreif** [comment]
After fixing the return vaue from plugins `/Plugin.Get` everything works fine, so the bug could be closed.

@cpuguy83  One question: in [your comment}(https://github.com/docker/docker/issues/21857#issuecomment-207057067) you said, `Get`was introduced to see, if this volume already exists. What was the motivation? Couldn't the plugin itself what todo if an existing volume is to be created?

**Comment by cpuguy83** [comment]
@holgerreif The reasoning is that volume names must currently be unique across all drivers, not just the docker host. So we check if a volume already exists with a given name, if so we check that it's the same driver and either reject or allow from there.

**Comment by holgerreif** [comment]
@cpuguy83 Thanks. Sounds sensible at first look and at least does not heart. As a side effect it buys some tolererance if the engine and volume driver plugin run alittle bit our of sync.

However, looking at the API calls I oberserve the follwing:
   1. First `docker volume create` : `Get` --> `Create` --> `Path` --> cli succeds
   2. Second `docker volume create` : `Path` --> cli succeds
For me, because volume is already created, the second `volume create` should fail.
Or there should be some documentation stating

_Creating a volume is an idempotent operation. Docker engine takes care for this, the plugin call is only issues for the first create_

For remove I observe something similar.
   1. First `docker volume rm` : `Get` (only for the volume driver providing the volume) -> `remove` --> cli succeds
   2. Second `docker volume rm` : `Get` (this time for all volume drivers) --> cli returns error
Again, this is robust. However it does not need urgently to be documented.


Back to volume creation. Since engine keeps track of volumes I see the following:

If I explicitely call `volume create` then for docker engine there are two cases:
   1. A volume with this name already exists --> engine has to reject request 
   2. A volume with this name does not exist yet --> volume driver could not have created one before, so `Get` always returns with empty response. 

The only scenario with a benefit for calling `Get` before `Create` is:
   * A volume with this name already exists, but engine does not know about it --> accept request and find out via `Get`, volume is created already, thus no need to call `Create`.

The only circumstances where this can happen is the restart of docker daemon while the volume driver runs and a volume is created. Is this a case where we expect a well definded result? And while we are at at: What behaviour do we expect if the volume driver dies and is restarted?

Two answers worth getting into same documentation...

**Comment by thaJeztah** [comment]
ping @cpuguy83

**Comment by cpuguy83** [comment]
The issue here is how `docker run -v` has been handled in the past (e.g. implicit create) along with race conditions inherent in checking for existence and then creating.
Maybe we can just split `Create` into `IdempotentCreate` and `Create` and be done with it.
